### PR TITLE
blockchain: Fix and enhance the blockchain iterator

### PR DESCRIPTION
**Pull request by g11tech**
Blockchain's reorg handling was incorrect as well as client execution's usage of iterator. Fixed those.

Also enhanced the iterator with the option to run callback outside the `runWithlock` as executing the block doesn't require one to lock the blockchain, which may very well be used to get/make blockchain updates while the block is executing. This may make concurrency more efficient.

### CODE REVIEW

**Comment by holgerd77** [comment]
Rebased this via UI.

File: packages/blockchain/test/iterator.spec.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
+import * as tape from 'tape'
+
+import { Blockchain } from '../src'
+
+import { createTestDB, generateBlockchain, generateConsecutiveBlock } from './util'
+
+import type { Block } from '@ethereumjs/block'
+
+tape('blockchain test', (t) => {
+  t.test('should iterate through 24 blocks without reorg', async (st) => {
```
Comment (holgerd77) [inline packages/blockchain/test/iterator.spec.ts:10]: Ah, great to have these tests extracted! 👍

**Review by holgerd77 — APPROVED** [review]
Nice, this LGTM!

Feel free to merge.

File: packages/blockchain/test/iterator.spec.ts (hunk 1 above)
Comment (g11tech) [reply packages/blockchain/test/iterator.spec.ts:10]: :slightly_smiling_face:  also enhanced the tests with reorg creation and checking

Comment (holgerd77) [reply packages/blockchain/test/iterator.spec.ts:10]: Ah sorry, didn't have such a close look and wasn't totally obvious through this extraction and the diff therefore not displaying, but, yes: that's great too for sure! 🙂

**Pull request merged**